### PR TITLE
feat(fqdn): json decode & encode for rpc_host_port

### PR DESCRIPTION
**Pull request by GehaFearless**
Issue: https://github.com/apache/incubator-pegasus/issues/1810

Implement json_decode & json_encode for class `rpc_host_port`.

### CODE REVIEW

File: src/utils/test/json_helper_test.cpp [hunk 1]
```diff
@@ -505,4 +506,14 @@ TEST(json_helper, upgrade_downgrade)
     ASSERT_EQ(n2.c, o.c);
 }
 
+TEST(json_helper, host_port_encode_decode)
+{
+    dsn::host_port hp("localhost", 8888);
+    dsn::blob bb = dsn::json::json_forwarder<decltype(hp)>::encode(hp);
+    dsn::host_port hp2;
+    dsn::json::json_forwarder<decltype(hp)>::decode(bb, hp2);
```
Comment (acelyc111) [inline src/utils/test/json_helper_test.cpp:514]:
```suggestion
    dsn::json::json_forwarder<dsn::host_port>::decode(bb, hp2);
```

File: src/runtime/rpc/rpc_host_port.cpp [hunk 2]
```diff
@@ -86,6 +87,34 @@ host_port::host_port(rpc_address addr)
...
+        return false;
+    }
+
+    struct addrinfo hints;
+    memset(&hints, 0, sizeof(hints));
+    hints.ai_family = AF_INET;
+    hints.ai_socktype = SOCK_STREAM;
+    AddrInfo result;
+    auto err_code = GetAddrInfo(_host, hints, &result);
+
```
Comment (acelyc111) [inline src/runtime/rpc/rpc_host_port.cpp:109]:
```suggestion
```

**Review by acelyc111 — APPROVED** [review]

**Review by acelyc111 — APPROVED** [review]

File: src/runtime/rpc/rpc_host_port.cpp [hunk 3]
```diff
@@ -86,6 +87,33 @@ host_port::host_port(rpc_address addr)
     _type = addr.type();
 }
 
+bool host_port::from_string(const std::string s)
```
Comment (empiredan) [inline src/runtime/rpc/rpc_host_port.cpp:90]:
```suggestion
bool host_port::from_string(const std::string &s)
```

File: src/runtime/rpc/rpc_host_port.cpp [hunk 4]
```diff
@@ -86,6 +87,33 @@ host_port::host_port(rpc_address addr)
...
+{
+    auto pos = s.find_last_of(':');
```
Comment (empiredan) [inline src/runtime/rpc/rpc_host_port.cpp:92]:
```suggestion
    const auto pos = s.find_last_of(':');
```

File: src/runtime/rpc/rpc_host_port.cpp [hunk 5]
```diff
@@ -86,6 +87,33 @@ host_port::host_port(rpc_address addr)
...
+    if (pos == std::string::npos) {
+        return false;
+    }
+    _host = s.substr(0, pos);
+    std::string port = s.substr(pos + 1);
+
+    if (!internal::buf2unsigned(port, _port) || _port > UINT16_MAX) {
```
Comment (empiredan) [inline src/runtime/rpc/rpc_host_port.cpp:99]: Consider add a new function `buf2uint16` ? In principle, `internal::buf2unsigned` is an internal function and should be used only in `src/utils/string_conv.h`.

On the other hand, there is no need to check the validity of `_port`, which has been done in `buf2unsigned`:
```suggestion
    if (!buf2uint16(port, _port)) {
```

File: src/runtime/rpc/rpc_host_port.cpp [hunk 6]
```diff
@@ -86,6 +87,33 @@ host_port::host_port(rpc_address addr)
...
+        return false;
+    }
+
+    struct addrinfo hints;
+    memset(&hints, 0, sizeof(hints));
+    hints.ai_family = AF_INET;
+    hints.ai_socktype = SOCK_STREAM;
+    AddrInfo result;
+    auto err_code = GetAddrInfo(_host, hints, &result);
+    if (dsn_unlikely(!err_code.is_ok())) {
```
Comment (empiredan) [inline src/runtime/rpc/rpc_host_port.cpp:109]:
```suggestion
    if (dsn_unlikely(!GetAddrInfo(_host, hints, &result))) {
```

File: src/runtime/test/host_port_test.cpp [hunk 7]
```diff
@@ -87,6 +87,16 @@ TEST(host_port_test, operators)
     ASSERT_NE(hp, hp2);
     ASSERT_FALSE(hp.is_invalid());
     ASSERT_TRUE(hp2.is_invalid());
+
+    std::string hp_str = "localhost:8080";
+    host_port hp3;
+    ASSERT_TRUE(hp3.from_string(hp_str));
+    ASSERT_EQ(hp, hp3);
```
Comment (empiredan) [inline src/runtime/test/host_port_test.cpp:94]:
```suggestion
    ASSERT_EQ(hp, hp3);
    ASSERT_FALSE(hp3.is_invalid());
```

File: src/runtime/test/host_port_test.cpp [hunk 8]
```diff
@@ -87,6 +87,16 @@ TEST(host_port_test, operators)
     ASSERT_NE(hp, hp2);
     ASSERT_FALSE(hp.is_invalid());
     ASSERT_TRUE(hp2.is_invalid());
+
+    std::string hp_str = "localhost:8080";
+    host_port hp3;
+    ASSERT_TRUE(hp3.from_string(hp_str));
```
Comment (empiredan) [inline src/runtime/test/host_port_test.cpp:93]:
```suggestion
    ASSERT_TRUE(hp3.is_invalid());
    ASSERT_TRUE(hp3.from_string(hp_str));
```

File: src/runtime/test/host_port_test.cpp [hunk 9]
```diff
@@ -87,6 +87,16 @@ TEST(host_port_test, operators)
...
+    ASSERT_EQ(hp, hp3);
+
+    host_port hp4;
+    std::string hp_str2 = "pegasus:8080";
+    ASSERT_FALSE(hp4.from_string(hp_str2));
```
Comment (empiredan) [inline src/runtime/test/host_port_test.cpp:98]:
```suggestion
    ASSERT_TRUE(hp4.is_invalid());
    ASSERT_FALSE(hp4.from_string(hp_str2));
```

File: src/runtime/rpc/rpc_host_port.cpp [hunk 10]
```diff
@@ -86,6 +87,32 @@ host_port::host_port(rpc_address addr)
...
+bool host_port::from_string(const std::string &s)
+{
+    const auto pos = s.find_last_of(':');
+    if (pos == std::string::npos) {
+        return false;
+    }
+    _host = s.substr(0, pos);
+    std::string port = s.substr(pos + 1);
+
+    if (!dsn::buf2uint16(port, _port)) {
```
Comment (empiredan) [inline src/runtime/rpc/rpc_host_port.cpp:99]:
```suggestion
    if (dsn_unlikely(!dsn::buf2uint16(port, _port))) {
```

File: src/runtime/rpc/rpc_host_port.cpp [hunk 11]
```diff
@@ -86,6 +87,32 @@ host_port::host_port(rpc_address addr)
     _type = addr.type();
 }
 
+bool host_port::from_string(const std::string &s)
+{
+    const auto pos = s.find_last_of(':');
+    if (pos == std::string::npos) {
```
Comment (empiredan) [inline src/runtime/rpc/rpc_host_port.cpp:93]:
```suggestion
    if (dsn_unlikely(pos == std::string::npos)) {
```

File: src/utils/string_conv.h [hunk 12]
```diff
@@ -124,6 +124,10 @@ inline bool buf2uint64(absl::string_view buf, uint64_t &result)
     return internal::buf2unsigned(buf, result);
 }
 
+inline bool buf2uint16(absl::string_view buf, uint16_t &result)
+{
+    return internal::buf2unsigned(buf, result);
+}
```
Comment (empiredan) [inline src/utils/string_conv.h:130]:
```suggestion
}

```

File: src/runtime/rpc/rpc_host_port.cpp [hunk 13]
```diff
@@ -86,6 +87,32 @@ host_port::host_port(rpc_address addr)
...
+    if (dsn_unlikely(!dsn::buf2uint16(port, _port))) {
+        return false;
+    }
+
+    struct addrinfo hints;
+    memset(&hints, 0, sizeof(hints));
+    hints.ai_family = AF_INET;
+    hints.ai_socktype = SOCK_STREAM;
+    AddrInfo result;
+    if (dsn_unlikely(!GetAddrInfo(_host, hints, &result))) {
```
Comment (empiredan) [inline src/runtime/rpc/rpc_host_port.cpp:108]:
```suggestion
    if (dsn_unlikely(!GetAddrInfo(_host, hints, nullptr))) {
```

**Review by empiredan — APPROVED** [review]

**Review by acelyc111 — APPROVED** [review]

**Pull request merged**